File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/GenericChestContainer.java
package generations.gg.generations.core.generationscore.common.world.container;

import generations.gg.generations.core.generationscore.common.world.container.slots.LockedSlot;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.Container;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.MenuType;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public class GenericChestContainer extends AbstractContainerMenu {
    private final GenericContainer container;
    private final int guiWidth;
    private final int guiHeight;
    private final int playerInventoryX;
    private final Inventory playerInventory;
    private final int lockedSlot;

    public GenericChestContainer(int containerId, Inventory playerInventory, FriendlyByteBuf buf) {
        this(GenerationsContainers.GENERIC.get(), containerId, playerInventory, buf);
    }

    public GenericChestContainer(MenuType<?> type, int containerId, Inventory playerInventory, FriendlyByteBuf buf) {
        this(type, containerId, playerInventory, new GenericContainer.SimpleGenericContainer(buf.readVarInt(), buf.readVarInt()), buf.readVarInt());
    }

    public GenericChestContainer(int containerId, Inventory playerInventory, GenericContainer container) {
        this(GenerationsContainers.GENERIC.get(), containerId, playerInventory, container, -1);
    }

    public GenericChestContainer(MenuType<?> type, int containerId, Inventory playerInventory, GenericContainer container, int lockedSlot) {
        super(type, containerId);

        this.container = container;
        this.playerInventory = playerInventory;
        this.lockedSlot = lockedSlot;

        this.guiWidth = 14 + getInventoryWidth() * 18;
        this.guiHeight = 110 + (getInventoryHeight() * 18);
        container.startOpen(playerInventory.player);

        populate(container, 8, 16, 0, getInventoryHeight(), getInventoryWidth());

        this.playerInventoryX = guiWidth/2 - 80;

        populate(playerInventory, playerInventoryX, guiHeight - 82, 1, 3, 9);
        populate(playerInventory, playerInventoryX, guiHeight - 24, 0, 1, 9);
    }

    private void populate(Container container, int x, int y, int startingRow, int rows, int columns) {
        for (int row = 0; row < rows; row++) {
            for (int column = 0; column < columns; column++) {
                var slot = column + (startingRow + row) * columns;
                var xSlot = x + column * 18;
                var ySlot = y + row * 18;

                this.addSlot(container, slot, xSlot, ySlot);
            }
        }
    }

    public void addSlot(Container container, int slot, int x, int y) {
        this.addSlot(getSlot(container, slot, x, y));
    }

    public Slot getSlot(Container container, int slot, int x, int y) {
        return lockedSlot == slot ? new LockedSlot(container, slot, x, y) : new Slot(container, slot, x, y);
    }

    @Override
    public @NotNull ItemStack quickMoveStack(@NotNull Player player, int index) {
        ItemStack itemStack = ItemStack.EMPTY;

        Slot slot = this.slots.get(index);

        if(slot.hasItem()) {
            ItemStack itemStack1 = slot.getItem();
            itemStack = itemStack1.copy();

            if(index < container.getContainerSize() ? !this.moveItemStackTo(itemStack1, container.getContainerSize(), this.slots.size(), true) : !this.moveItemStackTo(itemStack1, 0, container.getContainerSize(), false)) {
                return ItemStack.EMPTY;
            }

            if(itemStack1.isEmpty()) {
                slot.set(ItemStack.EMPTY);
            } else {
                slot.setChanged();
            }
        }

        return itemStack;
    }

    @Override
    public boolean stillValid(@NotNull Player player) {
        return this.container.stillValid(player);
    }

    public void removed(@NotNull Player player) {
        super.removed(player);
        this.container.stopOpen(player);
    }

    public int getInventoryWidth() {
        return container.getWidth();
    }

    public int getInventoryHeight() {
        return container.getHeight();
    }

    public int getGuiWidth() {
        return guiWidth;
    }

    public int getGuiHeight() {
        return guiHeight;
    }

    public int getPlayerInventoryX() {
        return playerInventoryX;
    }

    public Container getContainer() {
        return container;
    }

    public void save(Player player) {}

    public Inventory getPlayerInventory() {
        return playerInventory;
    }

    protected int getLocked() {
        return lockedSlot;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/GenericChestContainer.kt
package generations.gg.generations.core.generationscore.common.world.container

import earth.terrarium.common_storage_lib.item.impl.SimpleItemStorage
import earth.terrarium.common_storage_lib.item.impl.vanilla.WrappedVanillaContainer
import earth.terrarium.common_storage_lib.item.util.ItemStorageData
import earth.terrarium.common_storage_lib.resources.item.ItemResource
import earth.terrarium.common_storage_lib.storage.base.CommonStorage
import earth.terrarium.common_storage_lib.storage.base.UpdateManager
import earth.terrarium.common_storage_lib.storage.util.MenuStorageSlot
import generations.gg.generations.core.generationscore.common.world.container.slots.LockedSlot
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.item.ItemStack

open class GenericChestContainer<T>(
    containerId: Int,
    @JvmField val playerInventory: Inventory,
    private val container: T,
    val inventoryWidth: Int,
    val inventoryHeight: Int,
    protected val locked: Int = -1
) : AbstractContainerMenu(
        GenerationsContainers.GENERIC.get(),
        containerId
    ) where T: CommonStorage<ItemResource>, T: UpdateManager<ItemStorageData> {
    @JvmField
    val guiWidth: Int =  14 + this.inventoryWidth * 18
    @JvmField
    val guiHeight: Int = 110 + (inventoryHeight * 18)
    @JvmField
    val playerInventoryX: Int

    constructor(containerId: Int, playerInventory: Inventory, buf: FriendlyByteBuf) : this(
        containerId, playerInventory,
        SimpleItemStorage(buf.readVarInt()) as T,
        buf.readVarInt(),
        buf.readVarInt()
    )

    init {
//        container.startOpen(playerInventory.player)



        populate(container, 8, 16, 0, inventoryHeight, inventoryWidth)

        this.playerInventoryX = guiWidth / 2 - 80

        val playerStorage = WrappedVanillaContainer(playerInventory)

        populate(playerStorage, playerInventoryX, guiHeight - 82, 1, 3, 9)
        populate(playerStorage, playerInventoryX, guiHeight - 24, 0, 1, 9)
    }

    private fun <V> populate(container: V, x: Int, y: Int, startingRow: Int, rows: Int, columns: Int) where V : CommonStorage<ItemResource> {
        for (row in 0..<rows) {
            for (column in 0..<columns) {
                val slot = column + (startingRow + row) * columns
                val xSlot = x + column * 18
                val ySlot = y + row * 18

                this.addSlot(container, slot, xSlot, ySlot)
            }
        }
    }

    private fun addSlot(container: CommonStorage<ItemResource>, slot: Int, x: Int, y: Int) {
        this.addSlot(getSlot(container, slot, x, y))
    }

    open fun getSlot(container: CommonStorage<ItemResource>, slot: Int, x: Int, y: Int): MenuStorageSlot {
        return if (locked == slot) LockedSlot(container, slot, x, y) else MenuStorageSlot(container, slot, x, y)
    }

    override fun quickMoveStack(player: Player, index: Int): ItemStack {
        var itemStack = ItemStack.EMPTY

        val slot = slots[index]

        if (slot.hasItem()) {
            val itemStack1 = slot.item
            itemStack = itemStack1.copy()

            if (if (index < container.size()) !this.moveItemStackTo(
                    itemStack1, container.size(),
                    slots.size, true
                ) else !this.moveItemStackTo(itemStack1, 0, container.size(), false)
            ) {
                return ItemStack.EMPTY
            }

            if (itemStack1.isEmpty) {
                slot.set(ItemStack.EMPTY)
            } else {
                slot.setChanged()
            }
        }

        return itemStack
    }

    override fun stillValid(player: Player): Boolean {
        return true
    }

    override fun removed(player: Player) {
        super.removed(player)
//        container.stopOpen(player)
    }

    fun getContainer(): T = container

    open fun save(player: Player?) {}
}
